### PR TITLE
Add mesh loss and gan loss for human mesh estimation.

### PR DESCRIPTION
**Pull request by zengwang430521**
Add mesh loss and gan loss for human mesh estimation.

### CODE REVIEW

File: mmpose/models/losses/mesh_loss.py [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+    # Transform points
+    points = torch.einsum('bij,bkj->bki', rotation, points)
+    points = points + translation.unsqueeze(1)
+
+    # Apply perspective distortion
+    projected_points = points / points[:, :, -1].unsqueeze(-1)
+
+    # Apply camera intrinsics
+    projected_points = torch.einsum('bij,bkj->bki', K, projected_points)
+    projected_points = projected_points[:, :, :-1]
```
Comment (innerlee) [inline mmpose/models/losses/mesh_loss.py:44]: what size is the last dim of projected_points

File: mmpose/models/losses/mesh_loss.py [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+                                             joints_2d_visible)
+        losses['joints_2d_loss'] = loss_joints_2d * self.joints_2d_loss_weight
+
+        return losses
+
+
+@LOSSES.register_module()
+class GANLoss(nn.Module):
+    """Define GAN loss. Adapted from https://github.com/open-mmlab/mmediting
+    Original licence:  Copyright (c) MMEditing Authors, under the Apache
```
Comment (innerlee) [inline mmpose/models/losses/mesh_loss.py:251]: rm notice for mm-sereis

File: mmpose/models/losses/mesh_loss.py (hunk 1 above)
Comment (zengwang430521) [reply mmpose/models/losses/mesh_loss.py:44]: It's the z coordinates. In other words, it's the depth.

File: mmpose/models/losses/mesh_loss.py (hunk 2 above)
Comment (zengwang430521) [reply mmpose/models/losses/mesh_loss.py:251]: OK, get it.

File: mmpose/models/losses/mesh_loss.py (hunk 1 above)
Comment (innerlee) [reply mmpose/models/losses/mesh_loss.py:44]: i mean, in L24, it says the returned shape is `[B, N, 3]`. but i guess the `[:, :, :-1]` might return some shape of `B, N, 2`

File: mmpose/models/losses/mesh_loss.py [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+        focal_length (Tensor([B,])): Focal length
+        camera_center (Tensor([B, 2])): Camera center
+
+    Returns:
+        projected_points (Tensor([B, N, 3])): Projected 2D
+            points in image space.
+    """
+
+    batch_size = points.shape[0]
+    K = torch.zeros([batch_size, 3, 3], device=points.device)
```
Comment (innerlee) [inline mmpose/models/losses/mesh_loss.py:29]: since K is the same for all samples in a batch, maybe try this

```python
K = torch.zeros([3, 3], device=points.device)
...
projected_points = torch.einsum('ij,bkj->bki', K, projected_points)
```

File: mmpose/models/losses/mesh_loss.py [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+
+        # Compute 2D reprojection loss for the 2D joints
+        pred_camera = output['camera']
+        gt_joints_2d = target['joints_2d']
+        joints_2d_visible = target['joints_2d_visible']
+        pred_joints_2d = self.project_points(pred_joints_3d, pred_camera)
+
+        # Normalize keypoints to [-1,1]
+        pred_joints_2d = pred_joints_2d / (self.img_res / 2.)
+        gt_joints_2d = gt_joints_2d / (self.img_res / 2.) - 1
```
Comment (innerlee) [inline mmpose/models/losses/mesh_loss.py:240]: why computation of the above two lines are different?

File: mmpose/models/losses/mesh_loss.py [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+    """
+
+    def __init__(self,
+                 joints_2d_loss_weight,
+                 joints_3d_loss_weight,
+                 vertices_loss_weight,
+                 smpl_pose_loss_weight,
+                 smpl_beta_loss_weight,
+                 focal_length=5000,
+                 img_res=256):
```
Comment (innerlee) [inline mmpose/models/losses/mesh_loss.py:73]: should this be required? is it possible that a user used a larger resolution and forgot to specify this?

File: mmpose/models/losses/mesh_loss.py [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+        """
+        batch_size = points_3d.shape[0]
+        device = points_3d.device
+        cam_t = torch.stack([
+            camera[:, 1], camera[:, 2], 2 * self.focal_length /
+            (self.img_res * camera[:, 0] + 1e-9)
+        ],
+                            dim=-1)
+        camera_center = camera.new_zeros([batch_size, 2])
+        rot_t = torch.eye(
```
Comment (innerlee) [inline mmpose/models/losses/mesh_loss.py:174]: why rot_t is hard-coded as identity? if so, then why it is needed?

File: mmpose/models/losses/mesh_loss.py [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+            losses (dict): dict of losses.
+        """
+        losses = {}
+
+        # Per-vertex loss for the shape
+        pred_vertices = output['vertices']
+
+        gt_vertices = target['vertices']
+        has_smpl = target['has_smpl']
+        loss_vertices = self.vertices_loss(pred_vertices, gt_vertices,
```
Comment (innerlee) [inline mmpose/models/losses/mesh_loss.py:207]: the pronunciation of `vertices_loss` is strange. how about change it to `vertex_loss`

File: mmpose/models/losses/mesh_loss.py [hunk 8]
```diff
@@ -0,0 +1,337 @@
...
+            Note that loss_weight is only for generators; and it is always 1.0
+            for discriminators.
+    """
+
+    def __init__(self,
+                 gan_type,
+                 real_label_val=1.0,
+                 fake_label_val=0.0,
+                 loss_weight=1.0):
+        super(GANLoss, self).__init__()
```
Comment (innerlee) [inline mmpose/models/losses/mesh_loss.py:266]: super

File: mmpose/models/losses/mesh_loss.py (hunk 1 above)
Comment (zengwang430521) [reply mmpose/models/losses/mesh_loss.py:44]: Sorry, it's a typo.  The returned shape should be [B, N, 2]. I will fix it.

File: mmpose/models/losses/mesh_loss.py (hunk 3 above)
Comment (zengwang430521) [reply mmpose/models/losses/mesh_loss.py:29]: K can be different for samples in a batch, if the focal_length and camera_center of samples are different.
So I prefer to keep K in shape [batch_size, 3, 3].

File: mmpose/models/losses/mesh_loss.py (hunk 4 above)
Comment (zengwang430521) [reply mmpose/models/losses/mesh_loss.py:240]: The ordinate origin of ` gt_joints_2d `is the top left corner of the input image. But the ordinate origin of ` pred_joints_2d` is the center of the input image. So the computation is different.

Comment (innerlee) [reply mmpose/models/losses/mesh_loss.py:240]: okay, add this info as a comment line to help future readers

File: mmpose/models/losses/mesh_loss.py (hunk 3 above)
Comment (innerlee) [reply mmpose/models/losses/mesh_loss.py:29]: yeah you are right

File: mmpose/models/losses/mesh_loss.py (hunk 5 above)
Comment (zengwang430521) [reply mmpose/models/losses/mesh_loss.py:73]: OK, I wiil make the img_res required.

File: mmpose/models/losses/mesh_loss.py (hunk 6 above)
Comment (zengwang430521) [reply mmpose/models/losses/mesh_loss.py:174]: The `perspective_projection` is used to realize the projection of a perspective camera model, which includes the rotation of camera. 
However, HMR only uses a weak perspective camera model and it ignores the rotation of camera. So the `rot_t` is fixed as identity.

File: mmpose/models/losses/mesh_loss.py (hunk 7 above)
Comment (zengwang430521) [reply mmpose/models/losses/mesh_loss.py:207]: Ok , I will modify it.

File: mmpose/models/losses/mesh_loss.py (hunk 8 above)
Comment (zengwang430521) [reply mmpose/models/losses/mesh_loss.py:266]: OK , fixed

File: mmpose/models/losses/mesh_loss.py (hunk 6 above)
Comment (innerlee) [reply mmpose/models/losses/mesh_loss.py:174]: is there any model that uses "non-weak" camera model?

File: mmpose/models/losses/mesh_loss.py (hunk 4 above)
Comment (zengwang430521) [reply mmpose/models/losses/mesh_loss.py:238]: OK, I will add the comment

File: mmpose/models/losses/mesh_loss.py (hunk 6 above)
Comment (zengwang430521) [reply mmpose/models/losses/mesh_loss.py:173]: No, most models use the weak perspective projection.

Comment (innerlee) [reply mmpose/models/losses/mesh_loss.py:173]: okay, let's leave it here for future reference.

File: mmpose/models/losses/mesh_loss.py [hunk 9]
```diff
@@ -0,0 +1,339 @@
...
+        losses['joints_3d_loss'] = loss_joints_3d * self.joints_3d_loss_weight
+
+        # Compute 2D reprojection loss for the 2D joints
+        pred_camera = output['camera']
+        gt_joints_2d = target['joints_2d']
+        joints_2d_visible = target['joints_2d_visible']
+        pred_joints_2d = self.project_points(pred_joints_3d, pred_camera)
+
+        # Normalize keypoints to [-1,1]
+        # The ordinate origin of pred_joints_2d is
```
Comment (innerlee) [inline mmpose/models/losses/mesh_loss.py:237]: coordinate origin?

File: mmpose/models/losses/mesh_loss.py [hunk 10]
```diff
@@ -0,0 +1,339 @@
...
+        #  the center of the input image.
+        pred_joints_2d = pred_joints_2d / (self.img_res / 2.)
```
Comment (innerlee) [inline mmpose/models/losses/mesh_loss.py:239]: just curious, 
> The ordinate origin of pred_joints_2d is the center of the input image.
if img width is 3, then what is the center of the input image? 1 or 1.5?

* * *
0 1 2

if center = 1.5, then have you checked that this convention is consistently used everywhere?
if center = 1, then its distance to the right-most pixel is 2.

Comment (zengwang430521) [reply mmpose/models/losses/mesh_loss.py:239]: It should be 1.
Maybe, during the normalization, I should use `self.img_res - 1` to replace `self.img_res`.

File: mmpose/models/losses/mesh_loss.py (hunk 9 above)
Comment (zengwang430521) [reply mmpose/models/losses/mesh_loss.py]: I will fix it.

File: mmpose/models/losses/mesh_loss.py (hunk 10 above)
Comment (innerlee) [reply mmpose/models/losses/mesh_loss.py:239]: will you do this in a following pr, or hold this pr for a decision?

Comment (zengwang430521) [reply mmpose/models/losses/mesh_loss.py]: I have modified the codes in this PR.

**Review by innerlee — APPROVED** [review]

**Comment by innerlee** [comment]
Need a following pr to update the change log
